### ISSUE TITLE
Capturing game view

### ISSUE THREAD

**Issue by xiaozhikang0916** [RadarNyan/CefEOBrowser#4]
In Chrome, by right-clicking game view it has a "Copy Image" in the context menu. With such menu we can get a capture in original size and quality of the game even if you resize the view. So I thing it is a better way to capture a game view.

However, the most used way of getting image from a canvas, the `toDataURL` method is not working with Haedware Accelerating enabled in Chrome by returning a solid black image, and if such option is turned off, this method works well.

Do you have any idea?

Demo code is pushed  to [my frok project](https://github.com/xiaozhikang0916/CefEOBrowser/tree/capture) or you can try it in Chrome console

**Comment by silfumus** [comment]
For reference: https://github.com/Grabacr07/KanColleViewer/commit/f1b612a0a833c6df0cecaaa21ee18b87ec8f81ac

**Comment by xiaozhikang0916** [comment]
@silfumus So the reference commit is chaning the game's setting by replacing js code before game is loaded. But I don't think it is a good idea because currently I know nothing about the PIXI lib and may lead to some performance problem, since now performance IS a problem.

By the way, I tried disabling hardware accelerating by `settings.DisableGpuAcceleration();` and i can get the picture. But I still want to know if there is a way we can get the capture without disabling hardware accelerating nor changing game code, like what Chrome has done.

**Comment by silfumus** [comment]
The only alternative I can think of is to use the `CefSharp.Wpf` which has [Offscreen Rendering](https://github.com/cefsharp/CefSharp/wiki/General-Usage#offscreen-rendering-osr). That way, you could just grab the bitmap from its render buffer. But its performance is slower than `CefSharp.WinForms`.

Also, disabling `Hardware acceleration` might be good trade-off for some, since it's also required to avoid blank screen while streaming/recording KanColle.

**Comment by RadarNyan** [comment]
TL;DR, this is impossible due to the game is rendered with WebGL.

My first thought was to get the original size screenshot from the canvas, but canvas.ToDataUrl() would just return blank image, unless:

1) Disable GPU acceleration, which I have tried, results in losing some visual effects which I don't like (eg: destroyed ships remains in color instead of gray out)

2) Modify game code or game engine (pixi.js) code to set ```preserveDrawingBuffer = true``` (this property can't be set on-the-fly which means once enabled it can't be disabled, vice versa), which I also tried, it's not only very complicated to write in code, but also results in a performance drop (on my old laptop) which can't be justified by just having the ability to make screenshot in original size.

So I simply used ```CopyFromScreen``` to get what's on screen, results in only getting the "zoomed" screenshot. I don't know how Chrome pulls this off but it's not available from cef.

**Comment by RadarNyan** [comment]
Close due to project cancellation.
